### DISCUSSION TITLE
Connecting to different Targets

### DISCUSSION THREAD

**Discussion by ordinaryorange** [cefsharp/CefSharp.Dom#37] [General]
In devtools protocol a call to `Target.attachToTarget` will return a `sessionId` of the newly established session to given target.
"flat mode" can then be used to issue commands to any connected target session
e.g. the raw message will look like this
`{method: "Devtools.method", "params": {"method params"}, "sessionId": "123456789"}`

The Cefsharp.Dom API does not look like it support multiple sessions ?
Is there workaround to issue commands over different sessions ?
Is this on the radar for this lib ?

**Comment by amaitland** [comment]
> e.g. the raw message will look like this
{method: "Devtools.method", "params": {"method params"}, "sessionId": "123456789"}

CEF doesn't support passing in a session ID that I'm aware of. Commands are send directly to the CefBrowserHost.

https://github.com/chromiumembedded/cef/blob/b318adddacccffb74bac95e49ac31b606d8b813e/include/cef_browser.h#L575

> Is there workaround to issue commands over different sessions ?

I don't believe it's currently possible. By all means double check the CEF source.

**Reply by ordinaryorange** [reply]
I can only superficially look at the Cef source, I'm not literate in C. :(
But it looks like you are correct.

PuppeteerSharp has a [Targets()](https://github.com/hardkoded/puppeteer-sharp/blob/866a80258aefa79a8bc82fb3519abdec54c0cbac/lib/PuppeteerSharp/Browser.cs#L184)  and from there a new session can be created via [CreateCDPSessionAsync](https://github.com/hardkoded/puppeteer-sharp/blob/c6487063b6dbbd02a314acde788e73f58f764b11/lib/PuppeteerSharp/Target.cs#L182) so my hack at the moment is to boot CefSharp with arg "remote-debugging-port" then hook into that with PuppeteerSharp. Super ugly.

I note [this](https://github.com/chromiumembedded/cef/blob/b318adddacccffb74bac95e49ac31b606d8b813e/include/cef_browser.h#L559) method in Cef, but I cant find the equiv in CefSharp.Dom ? This might work as I could compose my own message with the `sessionId` ?

Thoughts ?

**Reply by amaitland** [reply]
> PuppeteerSharp has a [Targets()](https://github.com/hardkoded/puppeteer-sharp/blob/866a80258aefa79a8bc82fb3519abdec54c0cbac/lib/PuppeteerSharp/Browser.cs#L184) and from there a new session can be created via [CreateCDPSessionAsync](https://github.com/hardkoded/puppeteer-sharp/blob/c6487063b6dbbd02a314acde788e73f58f764b11/lib/PuppeteerSharp/Target.cs#L182) so my hack at the moment is to boot CefSharp with arg "remote-debugging-port" then hook into that with PuppeteerSharp. Super ugly.

What do you need to create a new session for exactly? 

> I note [this](https://github.com/chromiumembedded/cef/blob/b318adddacccffb74bac95e49ac31b606d8b813e/include/cef_browser.h#L559) method in Cef, but I cant find the equiv in CefSharp.Dom ? This might work as I could compose my own message with the sessionId ?

I see no reference to session in the doc. You can look at the source to see what it does or alternatively ask on https://magpcss.org/ceforum/index.php to see if CEF supports what you are after.

**Reply by ordinaryorange** [reply]
> What do you need to create a new session for exactly?

I have a webpage in Cefsharp that produces a popup, and I need to run some JS on the popup. I checked out if I could grab the window via a LifespanHandler, but discovered it is not possible to start a new Cef browser, block the native one and return the Cef window handle.

So I figured I could do a `Runtime.evaluate` on the popup. But the popup is a different Target which I need to connect to first.

> I see no reference to session in the doc. 

I don't think we need one, looking at that method it just takes a string, which I assume will be the full message e.g. `{method: "Devtools.method", "params": {"method params"}, "sessionId": "123456789"}`, I'll just include the session in the string. If I'm making the right assumptions here.... Can we do a `SendDevToolsMessage(string)` in CefSharp or Dom ?

**Reply by amaitland** [reply]
> I have a webpage in Cefsharp that produces a popup, and I need to run some JS on the popup. I checked out if I could grab the window via a LifespanHandler, but discovered it is not possible to start a new Cef browser, block the native one and return the Cef window handle.

Not quite sure what you mean by this. 

Creating a devtoolscontext for a popup should already be possible, just no extension method for it yet. You can implement your own though. You'll need a lifespanhandler. 

https://github.com/cefsharp/CefSharp.Dom/blob/main/lib/PuppeteerSharp/WebBrowserExtensions.cs#L58
http://cefsharp.github.io/api/104.4.x/html/M_CefSharp_IBrowser_GetHost.htm

**Reply by ordinaryorange** [reply]
I think we have crossed wires. I'll try and re explain.

I kick off a CefSharp browser and navigate to a website.
On that website is a button which I click, and a popup window appears. The popup occurs via the Window.open javascript function in the page. I note in the source js of the page that a reference to the WindowProxy object is maintained (`var winproxy = Window.open`), giving scoped access to the popup window for the rest of the calling code (which is important).

So I tried hooking up an [OnBeforPopup](https://cefsharp.github.io/api/57.0.0/html/M_CefSharp_ILifeSpanHandler_OnBeforePopup.htm) to try and intercept the new popup.

The `IWebBrowser` and `IBrowser` that are passed in refer to the original parent window. So any Devtools access via those references will only result in the CDP Target being the original host page. (and it is not possible to access the "native" popup window via that Target)

What I attempted to do was to start a new Cef browser with the associated `IPopupFeatures` and `IWindowInfo`  settings, which I can do, but then I need to assign that browser to the `out newBrowser` argument (so a window proxy might hopefully be returned to the calling js), and also return true to block the original native window. But this just throws a `NotSupportedException`. I did see [this issue](https://github.com/cefsharp/CefSharp/issues/3847) in case you ask.

So if I cant make a CefPopup (which would give me direct CDP access to that popup). Unless I've totally missed something here.

My current idea is to hook into the "native" popup via CDP by attaching to it via the `Target.attachToTarget` command. This will return a `sessionId` and from there I can then run any CDP method against that session (in what is called 'flat' mode) so long as the `sessionId` is included in the CDP message.

So in theory I could pass a manually constructed CDP message (which includes a sessionid) to the CefBrowser that invoked the 'native' popup, and the chromium infrastructure would effectively run that CDP message in the context of the popup.

The Cef method I referenced above says
>//Send a method call message over the DevTools protocol. |message| must be a
  // UTF8-encoded JSON dictionary that contains "id" (int), "method" (string)
  // and "params" (dictionary, optional) values. 

At a guess I reckon Cef wont touch the message and simply pass it onto the chromium infrastructure, and if the message has an additional sessionId it all might just work. If I had access to an equivalent method in CefSharp or Dom (and it worked), I could dump using PuppeteerSharp, but I cant seem to find such method.

Sorry, long winded, but just wanted to eliminate confusion.

**Reply by ordinaryorange** [reply]
Ha, I just found it. `browser.GetBrowserHost().SendDevToolsMessage()` going to try now....

**Reply by amaitland** [reply]
> The IWebBrowser and IBrowser that are passed in refer to the original parent window. So any Devtools access via those references will only result in the CDP Target being the original host page. (and it is not possible to access the "native" popup window via that Target)

That is before the popup is created as the name implies, you need OnAfterCreated to get a reference to the `IBrowser` instance associated with the popup. I'd suggest you try that without attempting any other changes in `LifeSpanHandler`.

> So if I cant make a CefPopup (which would give me direct CDP access to that popup). Unless I've totally missed something here.

What are you trying to do with the popup exactly? Are you using WinForms of WPF?

**Reply by ordinaryorange** [reply]
_Yep, I totally missed something here...._

You are correct. `OnAfterCreated` and the `IBrowser` are what I need. I completely overlooked it.
I just tested the CDP execution context on the `IBrowser` and it is working beautifully.
Your help has been invaluable, I should have asked this earlier and would have got 3 days of my life back...
You can definitely mark this one as answered!
